File: admin-fe/src/app/service/nhan-vien.service.ts
import { HttpClient } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";

import { PagedResponse } from "../model/interface/paged-response.interface";
import { NhanVienResponse } from "../model/interface/nhan-vien-response.interface";

@Injectable({
  providedIn: "root",
})
export class NhanVienService {
  private readonly apiUrl = "http://localhost:8080/nhan-vien";

  constructor(private http: HttpClient) {}

  public getAll(
    pageNumber: number = 1,
    pageSize: number = 5,
    search: string = ""
  ): Observable<PagedResponse<NhanVienResponse>> {
    const param = `?pageNumber=${pageNumber}&pageSize=${pageSize}&search=${search}`;
    return this.http.get<PagedResponse<NhanVienResponse>>(
      `${this.apiUrl}/get-all${param}`
    );
  }

  public getOneById(id: number): Observable<NhanVienResponse> {
    return this.http.get<NhanVienResponse>(
      `${this.apiUrl}/get-one-by-id/${id}`
    );
  }

  public add(
    nhanVien: NhanVienResponse,
    file: File
  ): Observable<NhanVienResponse> {
    const formData = new FormData();
    formData.append("request", JSON.stringify(nhanVien));
    formData.append("khachHangImage", file);
    return this.http.post<NhanVienResponse>(`${this.apiUrl}/add`, formData);
  }

  public delete(id: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/delete/${id}`);
  }

  public update(
    nhanVien: NhanVienResponse,
    id: number,
    file: File
  ): Observable<NhanVienResponse> {
    const formData = new FormData();
    formData.append("request", JSON.stringify(nhanVien));
    formData.append("khachHangImage", file);
    return this.http.put<NhanVienResponse>(
      `${this.apiUrl}/update/${id}`,
      formData
    );
  }

  public filter(
    pageNumber: number,
    pageSize: number,
    search: string,
    gioiTinhFilter: number[],
    trangThaiFilter: number[]
  ): Observable<PagedResponse<NhanVienResponse>> {
    const param = `?pageNumber=${pageNumber}&pageSize=${pageSize}&search=${search}&gioiTinhFilter=${gioiTinhFilter}&trangThaiFilter=${trangThaiFilter}`;
    return this.http.get<PagedResponse<NhanVienResponse>>(
      `${this.apiUrl}/filter${param}`
    );
  }
}

File: back-end/back-end/src/main/java/com/datn/backend/service/NhanVienService.java
package com.datn.backend.service;

import com.datn.backend.dto.request.AddNhanVienRequest;
import com.datn.backend.dto.response.NhanVienResponse;
import com.datn.backend.dto.response.PagedResponse;
import com.datn.backend.model.nhan_vien.NhanVien;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.List;

public interface NhanVienService {

    NhanVien add(AddNhanVienRequest request, MultipartFile multipartFile) throws IOException;

    PagedResponse<NhanVienResponse> getAll(int pageNumber, int pageSize, String search);

    NhanVienResponse getOneById(Integer id);

    NhanVien update(AddNhanVienRequest request, Integer id, MultipartFile multipartFile);

    NhanVien delete(Integer id);

    PagedResponse<NhanVienResponse> filter(int pageNumber, int pageSize, List<Integer> gioiTinhFilter, List<Integer> trangThaiFilter, String search);
}
